src/idp/response_builder.rs

```diff
@@ -0,0 +1,180 @@
+use crate::schema::{Response, Assertion, Issuer, Status, StatusCode, Subject, AuthnStatement, AuthnContext, AttributeStatement, SubjectConfirmation, SubjectConfirmationData, Conditions, AuthnContextClassRef, AudienceRestriction, SubjectNameID};
+use chrono::{Utc};
+use crate::signature::{Signature, SignedInfo, SignatureValue, CanonicalizationMethod, SignatureMethod, Reference, DigestMethod, DigestValue, Transform, Transforms};
+use crate::key_info::{KeyInfo, X509Data};
+use crate::attribute::{Attribute, AttributeValue};
+
+use crate::crypto;
+use super::sp_extractor::RequiredAttribute;
+
+fn signature_template(ref_id: &str, x509_cert_der: &[u8]) -> Signature {
+    Signature {
+        id: None,
+        signed_info: SignedInfo {
+            id: None,
+            canonicalization_method: CanonicalizationMethod {
+                algorithm: "http://www.w3.org/2001/10/xml-exc-c14n#".to_string()
+            },
+            signature_method: SignatureMethod {
+                algorithm: "http://www.w3.org/2001/04/xmldsig-more#rsa-sha256".to_string(), hmac_output_length: None
+            },
+            reference: vec![
+                Reference {
+                    transforms: Some(Transforms {
+                        transforms: vec![Transform {
+                            algorithm: "http://www.w3.org/2000/09/xmldsig#enveloped-signature".to_string(),
+                            xpath: None,
+                        }, Transform {
+                            algorithm: "http://www.w3.org/2001/10/xml-exc-c14n#".to_string(),
+                            xpath: None,
+                        }]
+                    }),
+                    digest_method: DigestMethod { algorithm: "http://www.w3.org/2000/09/xmldsig#sha1".to_string() },
+                    digest_value: DigestValue { base64_content: "".to_string() },
+                    uri: Some(format!("#{}", ref_id)),
+                    reference_type: None,
+                    id: None
+                }
+            ]
+        },
+        signature_value: SignatureValue { id: None, base64_content: "".to_string() },
+        key_info: Some(vec![KeyInfo {
+            id: None,
+            x509_data: Some(X509Data {
+                certificate: Some(crypto::mime_encode_x509_cert(x509_cert_der))
+            })
+        }]),
+    }
+}
+
+fn build_conditions(audience: &str) -> Conditions {
+    Conditions {
+        not_before: None, //Some(Utc::now().sub(Duration::minutes(1))),
+        not_on_or_after: None,//Some(Utc::now().add(Duration::days(3650))),
```

Comment:
Did you mean to uncomment these? Should probably have a default clock skew of `+-` 1 or 2 min.